### PR TITLE
Workaround test failure caused by clang, -ffast-math and santizers

### LINKED ISSUE #3077: Tests failure with clang + ubsan + RelWithDebInfo (low priority)
**Describe the bug**
Two tests fail consistently with a clang+ubsan+complex+RelWithDebInfo combo on Ubuntu 20.04.
Error is related to numerical discrepancy, not ubsan.
Low priority as it's not reproducible with any other combination.

**To Reproduce**
Steps to reproduce the behavior on Ubuntu 20.04 (using defaults):
1. current develop branch
2. Only reproducible with clang+ubsan+complex+RelWithDebInfo
3. Configuration:
```
CC=clang CXX=clang++ cmake -GNinja \
  -DCMAKE_BUILD_TYPE=RelWithDebInfo \
  -DMPI_C_COMPILER=mpicc -DMPI_CXX_COMPILER=mpicxx \
  -DENABLE_SANITIZER=ubsan \
  -DQMC_COMPLEX=1 \
  $QMCPACK_DIR
```
4. Build: `ninja`
5. Launch failed tests with numerical discrepancies:
  - `ctest -VV -R  deterministic-unit_test_wavefunction_trialwf`
  - `ctest -VV -R  deterministic-unit_test_wavefunction_sposet`

**Expected behavior**
Tests should pass as with any other compiler and settings combination.

**System:**
 - Ubuntu 20.04
 - Also seen in GitHub Actions container

**Additional context**
Hunches: race condition or some collateral bug triggered by optimizations (clang + LTO)

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

The combination of clang, `-ffast-math` AND one of the sanitizers enabled, is causing the test `deterministic-unit_test_wavefunction_trialwf` test to fail. 

```
21: ../src/QMCWaveFunctions/tests/test_TrialWaveFunction_diamondC_2x1x1.cpp:271: FAILED:
21:   REQUIRE( grad_old[0][0] == ComplexApprox(ValueType(713.71203320653, 0.020838031926441)).epsilon(8e-5) )
21: with expansion:
21:   (713.772,0.0527774)
21:   ==
21:   ComplexApprox( (713.712,0.020838) )
```

Removing the `-ffast-math` flag when building src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp is enough to make the tests pass.

Fixes #3077

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix

### Does this introduce a breaking change?

- No

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

**Comment by ye-luo** [comment]
Try changing to epsilon(1e-4)

**Review by prckent — CHANGES_REQUESTED** [review]
1. This is a non-trivial numerical error that we should make sure that we understand. Since this is in a unit test it should be practical to do so, and one of our reasons for having them is to catch things like this. The difference looks to be substantial at first glance, particularly for double precision. e.g. Perhaps we should be completely blocking fast-math with Clang and sanitizers enabled?  
2. We haven't worked out a formal policy around this, but I would prefer that for workarounds we set a flag in the compiler settings file and then test the flag. This would let us avoid the CMakefiles in src/* becoming littered with workarounds that are hard to keep track of. We are already doing this for missing &/or incorrect OpenMP target offload features.

@quantumsteve Please poke around in the unit test to try to identify what aspect is sensitive to fast-math. A good outcome would be a clang bug report.

**Comment by williamfgc** [comment]
FYI, I was playing with the debugger and I got to this point where differences start to show in the outputs of the `TrialWaveFunction::evalGrad` function [here](https://github.com/QMCPACK/qmcpack/blob/develop/src/QMCWaveFunctions/tests/test_TrialWaveFunction_diamondC_2x1x1.cpp#L265):

clang 10,11,12 + sanitizer + optimizer (failure)
```
evalGrad (713.77172813203,0.052777408712984) (713.77195321729,0.052777418805931) (-768.49254626048,-0.055357291163561) (118.02615857801,-0.0022475840711849) (118.02611412501,-0.0022475840711849) (-118.46286858168,0.0022414765057711)
```

correct (any other combo):
```
evalGrad (713.74580276991,0.052774339474592) (713.74575969796,0.052729498613329) (-768.46457083889,-0.055354036795209) (118.02569663111,-0.0022475751109497) (118.02565414806,-0.0022536826520141) (-118.46240652107,0.0022414675698894)
```

I couldn't backtrace further as there is plenty of optimized-out variables (as expected), so I might add print statements as we go. I agree with @prckent , finding the root-cause of these types of errors might also help upstream developers (in this case `clang`). Thanks @quantumsteve for the workaround.

**Comment by quantumsteve** [comment]
@prckent I'll try to create a minimal working example. 

I'm able to reproduce the test failure by removing the `-ffast-math` flag and only adding `-fassociative-math -fno-signed-zeros` to `src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp`. I hope we can get the example down to something small enough for an LLVM bug report.

**Comment by prckent** [comment]
@quantumsteve It might well be tied in with vectorization and the target architecture. Perhaps try a vanilla x86 target?

**Comment by quantumsteve** [comment]
> @quantumsteve It might well be tied in with vectorization and the target architecture. Perhaps try a vanilla x86 target?

The test passes if I remove `-march=native`.

The test passes if I replace `-march=native` with `-mavx2`.

The test fails if I replace `-march=native with `-mfma`. 

🤔

**Comment by prckent** [comment]
Hmm. Will be easier with a minimal example. Assuming that we don't somehow have a bug lurking (say, bad memory alignment), it does point to vectorization, fma etc.

**Comment by williamfgc** [comment]
I have a question/observation for @prckent @ye-luo @quantumsteve . The test is failing because of the [line](https://github.com/QMCPACK/qmcpack/blob/develop/src/QMCWaveFunctions/tests/test_TrialWaveFunction_diamondC_2x1x1.cpp#L271):

`REQUIRE(grad_old[0][0] == ComplexApprox(ValueType(713.71203320653, 0.020838031926441)).epsilon(8e-5));`

in which: 
- success: `grad_old[0][0] = (713.74580276991,0.052774339474592)`  
- failure: `grad_old[0][0] = (713.77172813203,0.052777408712984)`

The `Catch` comparison for a complex number is by magnitude, see [here](https://github.com/QMCPACK/qmcpack/blob/develop/external_codes/catch/complex_approx.hpp#L23).  Since the real part is dominant it seems to pass for a certain value. My question is if the imaginary part should also be close to the hardcoded value of `0.020838031926441` ? It seems off, so I wanted to bring it up.

**Comment by ye-luo** [comment]
@williamfgc the relative error is computed as the distance of two points in the complex plane divided by the module of the reference complex value. So when the real part is large, the absolute error on the imaginary part is relaxed as the modulus grows by the real part.

**Comment by prckent** [comment]
@williamfgc We care about both magnitude and phase/angle, or alternatively both individual components. Therefore both components are relevant and both are a little off.

**Comment by quantumsteve** [comment]
Passes with `-O0`, and fails with `-O1` and higher.

**Comment by quantumsteve** [comment]
I believe the test failure is caused by SplineC2C using `float` instead of `double` precision. I printed out the value of logdet/psiM right after being calculated and the differences are in about the 7th decimal place. 

https://github.com/QMCPACK/qmcpack/blob/5f40992c0d9038afc7f36e2490ab5316cd7d4552/src/QMCWaveFunctions/BsplineFactory/BsplineSet.h#L139
failed: (address sanitizers enabled, -O1)
```
20: psiM(bspline): (0.5456642508506775,0.5456642508506775) (0.6518307328224182,0.6518306136131287) (0.8073326945304871,0.8073326945304871) (0.9624435901641846,0.9624435901641846)
20: psiM(bspline): (0.8091580867767334,0.8091580867767334) (1.007317066192627,1.007317543029785) (0.8073327541351318,0.8073327541351318) (0.9624437689781189,0.9624435901641846)
20: psiM(bspline): (0.5456642508506775,0.5456642508506775) (0.6518307328224182,0.6518306136131287) (0.8073326945304871,0.8073326945304871) (0.9624435901641846,0.9624435901641846)
20: psiM(bspline): (0.4420452117919922,0.4420452117919922) (0.5367334485054016,0.5367334485054016) (0.8073326945304871,0.8073326945304871) (0.9624435901641846,0.9624435901641846)
20: psiM(bspline): (0.8091580867767334,0.8091580867767334) (1.007317066192627,1.007317543029785) (0.8073327541351318,0.8073327541351318) (0.9624437689781189,0.9624435901641846)
20: psiM(bspline): (0.8091580867767334,0.8091580867767334) (1.007317066192627,1.007317543029785) (0.8073327541351318,0.8073327541351318) (0.9624437689781189,0.9624435901641846)
```

passed: (address sanitizers enabled, -O0)
```
20: psiM(bspline): (0.5456642508506775,0.5456642508506775) (0.6518307328224182,0.6518306732177734) (0.8073327541351318,0.8073327541351318) (0.9624437093734741,0.9624435901641846)
20: psiM(bspline): (0.8091580867767334,0.8091580867767334) (1.007317066192627,1.007317423820496) (0.8073327541351318,0.8073327541351318) (0.9624437689781189,0.9624436497688293)
20: psiM(bspline): (0.5456642508506775,0.5456642508506775) (0.6518307328224182,0.6518306732177734) (0.8073327541351318,0.8073327541351318) (0.9624437093734741,0.9624435901641846)
20: psiM(bspline): (0.4420452117919922,0.4420452117919922) (0.5367334485054016,0.5367334485054016) (0.8073327541351318,0.8073327541351318) (0.9624437093734741,0.9624435901641846)
20: psiM(bspline): (0.8091580867767334,0.8091580867767334) (1.007317066192627,1.007317423820496) (0.8073327541351318,0.8073327541351318) (0.9624437689781189,0.9624436497688293)
20: psiM(bspline): (0.8091580867767334,0.8091580867767334) (1.007317066192627,1.007317423820496) (0.8073327541351318,0.8073327541351318) (0.9624437689781189,0.9624436497688293)
```

**Comment by quantumsteve** [comment]
Immediately after the recompute, `psiM_temp` is inverted. Would this amplify small differences in floating point values?

https://github.com/QMCPACK/qmcpack/blob/5f40992c0d9038afc7f36e2490ab5316cd7d4552/src/QMCWaveFunctions/Fermion/DiracDeterminant.cpp#L678

**Comment by prckent** [comment]
@quantumsteve I agree that matrix inversions are very sensitive to precision, so upgrading the precision might be a good+effective workaround for this test.

**Comment by quantumsteve** [comment]
double precision, address sanitizer (pass)
```
evalGrad (713.71203320677,0.020838031923745) (713.71203320678,0.020838031929145) (-768.42842826915,-0.020838032015719) (118.02653358655,-0.0022419843513362) (118.02653358655,-0.002241984350583) (-118.46325895635,0.0022419843501893)
```
double precision, no sanitizer (pass)
```
evalGrad (713.71203320671,0.02083803186462) (713.71203320672,0.020838031869574) (-768.42842826908,-0.020838031951865) (118.02653358655,-0.0022419843520401) (118.02653358655,-0.0022419843511556) (-118.46325895634,0.0022419843507426)
```

single precision, address sanitizer (fail)
```
evalGrad (713.77172813203,0.052777408712984) (713.77195321729,0.052777418805931) (-768.49254626048,-0.055357291163561) (118.02615857801,-0.0022475840711849) (118.02611412501,-0.0022475840711849) (-118.46286858168,0.0022414765057711)
```
single precision, no sanitizer (pass)
```
evalGrad (713.74580276991,0.052774339474592) (713.74575969796,0.052729498613329) (-768.46457083889,-0.055354036795209) (118.02569663111,-0.0022475751109497) (118.02565414806,-0.0022536826520141) (-118.46240652107,0.0022414675698894)
```

**Comment by ye-luo** [comment]
I got a similar issue on #3300
https://jenkins-ci.cels.anl.gov/job/QMCPACK-CI-GNU-MKL18/3006/label=compute004.mcs.anl.gov,type=Complex-noMPI/console
```
  (713.777,0.0155067)
  ==
  ComplexApprox( (713.712,0.020838) )
```
change the epsilon from 8e-5 to 1e-4 will be the simplest fix.

**Comment by prckent** [comment]
Better would be unit tests with both precisions and appropriate tolerances on the tests.

File: src/QMCWaveFunctions/tests/test_TrialWaveFunction_diamondC_2x1x1.cpp [hunk 1]
```diff
@@ -268,21 +286,20 @@ void testTrialWaveFunction_diamondC_2x1x1(const int ndelay)
...
-  REQUIRE(grad_old[1][1] == ComplexApprox(ValueType(118.02653358655, -0.0022419843498631)));
-  REQUIRE(grad_old[1][2] == ComplexApprox(ValueType(-118.46325895634, 0.0022419843493758)));
+  REQUIRE(grad_old[0][0] == ComplexApprox(ValueType(713.71203320653, 0.020838031926441)).epsilon(precision));
+  REQUIRE(grad_old[0][1] == ComplexApprox(ValueType(713.71203320654, 0.020838031928415)).epsilon(precision));
+  REQUIRE(grad_old[0][2] == ComplexApprox(ValueType(-768.42842826889, -0.020838032018344)).epsilon(precision));
+  REQUIRE(grad_old[1][0] == ComplexApprox(ValueType(118.02653358655, -0.0022419843505538)).epsilon(precision));
+  REQUIRE(grad_old[1][1] == ComplexApprox(ValueType(118.02653358655, -0.0022419843498631)).epsilon(precision));
+  REQUIRE(grad_old[1][2] == ComplexApprox(ValueType(-118.46325895634, 0.0022419843493758)).epsilon(precision));
 #else
-  REQUIRE(grad_old[0][0] == Approx(713.69119517454).epsilon(2e-4));
```
Comment (ye-luo) [inline src/QMCWaveFunctions/tests/test_TrialWaveFunction_diamondC_2x1x1.cpp:278]: this one was larger than 1e-4

**Review by ye-luo — COMMENTED** [review]
double tag is ignored in a mixed precision

3 cases in total. full precision build float and double precision spline. mixed precision build, float precision spline.

**Comment by ye-luo** [comment]
Test this please

**Comment by ye-luo** [comment]
Test this please

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]
LGTM

**Comment by ye-luo** [comment]
@prckent good to go?

**Review by prckent — APPROVED** [review]
I did not expect to see `<regex>`! No objections since it works well and avoids a full duplicate of the relevant XML block.

**Pull request merged**